src/cluster/mcl.cpp

```diff
@@ -265,42 +250,69 @@ void MCL::markov_process(Eigen::MatrixXf* m, float inflation, float expansion, u
+void MCL::print_stats(uint64_t nElements, uint32_t nComponents, uint32_t nComponentsLt1, vector<uint32_t>& sort_order, vector<vector<uint32_t>>& indices, SparseMatrixStream<float>* ms){
+	task_timer timer;
+	timer.go("Collecting stats");
 	const uint32_t chunk_size = config.cluster_mcl_chunk_size;
 	const uint32_t nThreads = min(config.threads_, nComponents / chunk_size);
 	const float inflation = (float) config.cluster_mcl_inflation;
 	const float expansion = (float) config.cluster_mcl_expansion;
+
 	vector<float> sparsities(nComponentsLt1);
-	uint64_t icomp = 0;
-	generate(sparsities.begin(), sparsities.end(), [&icomp, &sort_order, &indices, &components](){ 
-				const uint32_t iComponent = sort_order[icomp++];
-				const uint32_t ssize = components[iComponent].size();
-				const uint64_t dsize = indices[iComponent].size() * indices[iComponent].size();
-				return 1.0-(1.0 * ssize) / dsize;
-			});
 	vector<float> neighbors(nComponentsLt1);
-	icomp = 0;
-	generate(neighbors.begin(), neighbors.end(), [&icomp, &sort_order, &indices, &components](){ 
-				const uint32_t iComponent = sort_order[icomp++];
-				uint32_t neighbors = 0;
-				for(auto t: components[iComponent]){
-					if(t.row() != t.col()) neighbors++;
-				}
-				return neighbors / ((float) indices.size());
-			});
-
-	// rough memory calc before sorting sparsities
 	vector<float> memories(nComponentsLt1);
-	icomp = 0;
-	generate(memories.begin(), memories.end(), [&](){ 
-				const uint32_t iComponent = sort_order[icomp++];
-				if(sparsities[icomp - 1] >= config.cluster_mcl_sparsity_switch){ 
-					return indices[iComponent].size() * (1+pow(neighbors[icomp], expansion)) * (2 * sizeof(uint32_t) + sizeof(float));
-				}
-				else{
-					return (float) sizeof(float) * indices[iComponent].size() * indices[iComponent].size();
-				}
-			});
+
+	uint32_t max_job_size = 0;
+	for(uint32_t i=0; i<chunk_size; i++) max_job_size+=indices[sort_order[i]].size();
+	uint32_t my_counter = 0;
+	uint32_t component_counter = 1;
+
+	uint32_t my_chunk_size = chunk_size;
+	// Memory conservative collection of info from file
+	ms->allocate_read_buffer(1);
+	while(my_counter < nComponentsLt1){
+		unordered_set<uint32_t> all;
+		uint32_t upper_limit =  min(my_counter+my_chunk_size, nComponentsLt1);
+		vector<vector<uint32_t>*> loc_i;
+		for(uint32_t chunk_counter = my_counter; chunk_counter<upper_limit; chunk_counter++){
+			loc_i.push_back(&indices[sort_order[chunk_counter]]);
+		}
+		vector<vector<Eigen::Triplet<float>>> loc_c = ms->collect_components(&loc_i, 0);
+		uint32_t ichunk = 0;
+		generate(sparsities.begin()+my_counter, sparsities.begin()+upper_limit, [my_counter, ichunk, &sort_order, &indices, &loc_c]() mutable { 
+					const uint32_t iComponent = sort_order[my_counter++];
+					printf("ichunk %u\n",ichunk);
```

Comment:
Please write to stderr since stdout is reserved for output.

Reply:
This should not have been there at all, I will remove it. Good catch.

Reply:
Hm, I am pretty sure I had removed this prior to the commit. I am wondering whether there is a bug in github's diff or whether the removal is actually missing. We need to double-check!